File: tui/src/ws/protocol.rs
use serde::{Deserialize, Serialize};

// Client hello handshake message
#[derive(Debug, Clone, Serialize)]
pub struct ClientHello {
    #[serde(rename = "type")]
    pub msg_type: String, // "hello"
    #[serde(rename = "clientType")]
    pub client_type: String, // "tui"
    #[serde(rename = "clientId")]
    pub client_id: String,
    pub version: String,
}

impl ClientHello {
    pub fn new(client_id: String) -> Self {
        Self {
            msg_type: "hello".to_string(),
            client_type: "tui".to_string(),
            client_id,
            version: "0.1.0".to_string(),
        }
    }
}

// Server hello response
#[derive(Debug, Clone, Deserialize)]
pub struct ServerHello {
    #[serde(rename = "daemonVersion")]
    pub daemon_version: String,
    pub sessions: Vec<SessionSummary>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SessionSummary {
    pub id: String,
    pub name: String,
    #[serde(rename = "projectPath")]
    pub project_path: String,
    #[serde(rename = "appStatus")]
    pub app_status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MonitoringEvent {
    pub ts: String,
    pub source: String,
    /// Event type - daemon sends as "eventType", some legacy sources use "type"
    #[serde(alias = "type", rename = "eventType")]
    pub event_type: String,
    #[serde(default)]
    pub payload: serde_json::Value,
}

/// Session info matching SessionInfo from the server.
/// Represents a session that binds a project directory to a dedicated agent.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Session {
    pub id: String,
    pub name: String,
    #[serde(rename = "projectPath")]
    pub project_path: String,
    /// App status: "not_running", "starting", "running", "stopped", "error"
    #[serde(rename = "appStatus")]
    pub app_status: String,
    #[serde(rename = "vmServiceUri", default)]
    pub vm_service_uri: Option<String>,
    #[serde(default)]
    pub pid: Option<u32>,
    #[serde(rename = "connectedClients", default)]
    pub connected_clients: Vec<String>,
    #[serde(rename = "createdAt")]
    pub created_at: String,
    #[serde(rename = "lastActiveAt")]
    pub last_active_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum OutgoingMessage {
    Command {
        id: String,
        #[serde(rename = "clientId")]
        client_id: String,
        action: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        key: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        data: Option<serde_json::Value>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum IncomingMessage {
    CommandResponse(CommandResponse),
    AgentResponse(AgentResponse),
    /// Monitoring events from daemon (type: "event")
    Event(MonitoringEvent),
    /// Streaming agent events (type: "agent_stream")
    AgentStream(AgentStreamEvent),
}

/// Streaming event from agent execution
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentStreamEvent {
    pub id: String,
    #[serde(rename = "sessionId")]
    pub session_id: String,
    pub event: AgentEventKind,
}

/// Types of streaming events
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum AgentEventKind {
    TextDelta { text: String },
    ToolCallStart { 
        #[serde(rename = "toolName")]
        tool_name: String, 
        #[serde(rename = "toolCallId")]
        tool_call_id: String 
    },
    ToolCallEnd { 
        #[serde(rename = "toolName")]
        tool_name: String, 
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        result: Option<String>,
    },
    MessageComplete,  // Streaming finished for current message (immediate feedback)
    TaskComplete { summary: String },  // Full task/turn complete
    Error { message: String },
    /// User message from another client (e.g., MCP/Amp)
    UserMessage {
        text: String,
        #[serde(rename = "clientId")]
        client_id: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandResponse {
    pub id: String,
    pub success: bool,
    #[serde(default)]
    pub data: serde_json::Value,
    #[serde(default)]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentResponse {
    pub id: String,
    pub status: AgentStatus,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub question: Option<String>,
    /// Claude SDK session ID (managed by daemon, not used by TUI)
    #[serde(rename = "sdkSessionId", default)]
    pub sdk_session_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentStatus {
    Success,
    NeedsContext,
    Error,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_command_response() {
        let json = r#"{"type":"command_response","id":"123","success":true,"data":{"session":{"id":"sess-1","name":"test","projectPath":"/path","appStatus":"not_running","createdAt":"2024-01-01T00:00:00Z","lastActiveAt":"2024-01-01T00:00:00Z"}}}"#;
        let msg: IncomingMessage = serde_json::from_str(json).expect("parse failed");
        match msg {
            IncomingMessage::CommandResponse(resp) => {
                assert!(resp.success);
                assert_eq!(resp.id, "123");
                assert!(resp.data.get("session").is_some());
            }
            _ => panic!("Expected CommandResponse, got {:?}", msg),
        }
    }

    #[test]
    fn test_parse_session_from_response() {
        let json = r#"{"type":"command_response","id":"123","success":true,"data":{"session":{"id":"sess-1","name":"test","projectPath":"/path","appStatus":"not_running","createdAt":"2024-01-01T00:00:00Z","lastActiveAt":"2024-01-01T00:00:00Z"}}}"#;
        let msg: IncomingMessage = serde_json::from_str(json).expect("parse failed");
        if let IncomingMessage::CommandResponse(resp) = msg {
            let session_value = resp.data.get("session").expect("no session");
            let session: Session = serde_json::from_value(session_value.clone()).expect("parse session failed");
            assert_eq!(session.id, "sess-1");
            assert_eq!(session.name, "test");
            assert_eq!(session.project_path, "/path");
            assert_eq!(session.app_status, "not_running");
        }
    }

    #[test]
    fn test_parse_sessions_list() {
        let json = r#"{"type":"command_response","id":"123","success":true,"data":{"sessions":[{"id":"sess-1","name":"test","projectPath":"/path","appStatus":"running","vmServiceUri":"ws://127.0.0.1:5678","pid":1234,"createdAt":"2024-01-01T00:00:00Z","lastActiveAt":"2024-01-01T00:00:00Z"}]}}"#;
        let msg: IncomingMessage = serde_json::from_str(json).expect("parse failed");
        if let IncomingMessage::CommandResponse(resp) = msg {
            let sessions_arr = resp.data.get("sessions").and_then(|s| s.as_array()).expect("no sessions");
            let sessions: Vec<Session> = serde_json::from_value(serde_json::Value::Array(sessions_arr.clone())).expect("parse sessions failed");
            assert_eq!(sessions.len(), 1);
            assert_eq!(sessions[0].id, "sess-1");
            assert_eq!(sessions[0].pid, Some(1234));
        }
    }

    #[test]
    fn test_serialize_create_session_command() {
        let msg = OutgoingMessage::Command {
            id: "123".to_string(),
            client_id: "client-456".to_string(),
            action: "create_session".to_string(),
            key: None,
            data: Some(serde_json::json!({
                "name": "my-session",
                "projectPath": "/path/to/project",
            })),
        };
        let json = serde_json::to_string(&msg).expect("serialize failed");
        assert!(json.contains(r#""type":"command""#));
        assert!(json.contains(r#""clientId":"client-456""#));
        assert!(json.contains(r#""action":"create_session""#));
        assert!(json.contains(r#""name":"my-session""#));
        assert!(json.contains(r#""projectPath":"/path/to/project""#));
    }

    #[test]
    fn test_monitoring_event_fallback() {
        // Events with eventType field should parse as Event
        let json = r#"{"ts":"2024-01-01T00:00:00Z","source":"flutter","type":"event","eventType":"flutter.log","payload":{"line":"hello"}}"#;
        let msg: IncomingMessage = serde_json::from_str(json).expect("parse failed");
        match msg {
            IncomingMessage::Event(event) => {
                assert_eq!(event.source, "flutter");
                assert_eq!(event.event_type, "flutter.log");
            }
            _ => panic!("Expected Event, got {:?}", msg),
        }
    }

    #[test]
    fn test_session_destroyed_event() {
        let json = r#"{"ts":"2024-01-01T00:00:00Z","source":"session","type":"event","eventType":"session.destroyed","payload":{"sessionId":"sess-123"}}"#;
        let msg: IncomingMessage = serde_json::from_str(json).expect("parse failed");
        match msg {
            IncomingMessage::Event(event) => {
                assert_eq!(event.source, "session");
                assert_eq!(event.event_type, "session.destroyed");
                assert_eq!(event.payload["sessionId"], "sess-123");
            }
            _ => panic!("Expected Event, got {:?}", msg),
        }
    }
}

impl AgentResponse {
    pub fn to_monitoring_event(&self) -> MonitoringEvent {
        let (event_type, payload) = match self.status {
            AgentStatus::Success => (
                "agent_success".to_string(),
                serde_json::json!({
                    "summary": self.summary,
                }),
            ),
            AgentStatus::NeedsContext => (
                "agent_needs_context".to_string(),
                serde_json::json!({
                    "question": self.question,
                }),
            ),
            AgentStatus::Error => (
                "agent_error".to_string(),
                serde_json::json!({
                    "error": self.summary,
                }),
            ),
        };

        MonitoringEvent {
            ts: chrono::Utc::now().to_rfc3339(),
            source: "agent".to_string(),
            event_type,
            payload,
        }
    }
}

File: packages/fleeter-daemon/src/ws/protocol.ts
/**
 * WebSocket protocol types for fleeter-daemon.
 */

// ─────────────────────────────────────────────────────────────────────────────
// Client Types
// ─────────────────────────────────────────────────────────────────────────────

export type ClientType = 'mcp' | 'tui';

export interface ConnectedClient {
  id: string;
  type: ClientType;
  connectedAt: Date;
  currentSessionId: string | null;
}

// ─────────────────────────────────────────────────────────────────────────────
// Client → Daemon Messages
// ─────────────────────────────────────────────────────────────────────────────

export interface ClientHello {
  type: 'hello';
  clientType: ClientType;
  clientId: string;
  version: string;
}

export interface CommandMessage {
  type: 'command';
  id: string;
  clientId: string;
  action: CommandAction;
  data?: Record<string, unknown>;
}

export type CommandAction =
  // Session management
  | 'create_session'
  | 'destroy_session'
  | 'list_sessions'
  | 'connect_session'
  | 'disconnect_session'
  // App lifecycle
  | 'run_app'
  | 'stop_app'
  | 'hot_reload'
  | 'hot_restart'
  // Interaction tree
  | 'get_tree'
  | 'execute_interaction'
  | 'get_state'
  | 'batch'
  // Logs & errors
  | 'get_logs'
  | 'get_errors'
  // Context collection
  | 'get_context'
  // Agent
  | 'agent_message'
  | 'debug_agent_message'  // Debug agent (runtime-only investigation)
  // Status
  | 'get_status'
  | 'health_check';

export interface AgentToolCall {
  type: 'agent_tool_call';
  id: string;
  clientId: string;
  sessionId: string;
  intent: string;
  context?: string[];
  conversationId?: string;
}

export type ClientMessage = ClientHello | CommandMessage | AgentToolCall;

// ─────────────────────────────────────────────────────────────────────────────
// Daemon → Client Messages
// ─────────────────────────────────────────────────────────────────────────────

export interface ServerHello {
  type: 'hello_ack';
  daemonVersion: string;
  sessions: Array<{
    id: string;
    name: string;
    projectPath: string;
    appStatus: string;
  }>;
}

export interface CommandResponse {
  type: 'command_response';
  id: string;
  success: boolean;
  data?: unknown;
  error?: string;
}

export interface AgentResponse {
  type: 'agent_response';
  id: string;
  status: 'success' | 'failed' | 'needs_context';
  summary?: string;
  error?: string;
  question?: string;
  /** Claude SDK session ID for resumption (managed internally by daemon) */
  sdkSessionId?: string;
}

export interface MonitoringEvent {
  type: 'event';
  ts: string;
  source: 'flutter' | 'vm' | 'agent' | 'daemon' | 'session' | 'tree' | 'interaction';
  eventType: string;
  sessionId?: string;
  payload: unknown;
}

// ─────────────────────────────────────────────────────────────────────────────
// Agent Streaming Events
// ─────────────────────────────────────────────────────────────────────────────

export type AgentEventType =
  | { kind: 'text_delta'; text: string }
  | { kind: 'tool_call_start'; toolName: string; toolCallId: string }
  | { kind: 'tool_call_end'; toolName: string; toolCallId: string; result?: string }
  | { kind: 'message_complete' }  // Streaming finished for current message (immediate UI feedback)
  | { kind: 'task_complete'; summary: string }  // Full task/turn complete (after SDK finishes)
  | { kind: 'error'; message: string }
  | { kind: 'user_message'; text: string; clientId: string };  // User message from another client (e.g., MCP)

export interface AgentStreamEvent {
  type: 'agent_stream';
  id: string;
  sessionId: string;
  event: AgentEventType;
}

export type ServerMessage = ServerHello | CommandResponse | AgentResponse | MonitoringEvent | AgentStreamEvent;

// ─────────────────────────────────────────────────────────────────────────────
// Type Guards
// ─────────────────────────────────────────────────────────────────────────────

export function isClientHello(msg: unknown): msg is ClientHello {
  return typeof msg === 'object' && msg !== null && (msg as ClientHello).type === 'hello';
}

export function isCommandMessage(msg: unknown): msg is CommandMessage {
  return typeof msg === 'object' && msg !== null && (msg as CommandMessage).type === 'command';
}

export function isAgentToolCall(msg: unknown): msg is AgentToolCall {
  return typeof msg === 'object' && msg !== null && (msg as AgentToolCall).type === 'agent_tool_call';
}
